BLOCKING: Consensus: Move blocksize and related parameters to consensusparams ...without removing consensus/consensus.h [#6526 alternative]

jtimon opened the pull request:
This is a replacement for #6526 with my nits integrated in the commits where they were more relevant.
It has a smaller diff (138 additions and 65 deletions VS +116 −95) is more amenable to any block size change proposal (many of them will need a function), to #6382 (adding new test chains for block size tests) and to libconsenusus-related #6591 (which is also part of the process of making chainparams less global #5970).
I will rebase #6382, #6591 and #5970 on top of this. Please, @theuni close #6526 in favor of this if it passes your review.

Blocking:
- Consensus: Adapt declarations of most obviously consensus functions #6591 [jtimon/consensus-params-0.12.99]
- Chainparams cleanup #5970 [chainparams_cleanup](depends on #6163)
- Chainparams: Introduce std::numeric_limits<uint64_t>::max() new testnet chains to test different block sizes #6382 [chainparams-sizetest-0.11.99]

theuni committed:
consensus: don't define MAX_STANDARD_TX_SIGOPS in terms of block size

theuni committed:
consensus: teach CheckTransaction to check for an arbitrary max tx size

This is a no-op change.

Tests use a value of std::numeric_limits<uint64_t>::max() where necessary, to ensure that they're never
rejected when size isn't being tested.

theuni committed:
consensus: Move consensus constants into Consensus::Params and consensus.h (as functions)

The following are now tied to a chain rather than being defined as global
constants. Their values have not changed.

nMinTxSize
nMaxBlockSize
nMaxTxSize
nMaxBlockSigops
nCoinbaseMaturity

Also, for free (diff-wise):

Blocksize: Turn MAX_BLOCK_SIZE (nMaxBlockSize) and MAX_BLOCK_SIGOPS (nMaxBlockSigops) into functions

...which take Consensus::Params as parameter
This will be convenient to reduce the diff of any proposal that changes the blocksize as a hardfork

theuni committed:
consensus: teach ExtractMatches to check for an arbitrary max transaction number

This is a no-op change. For now, everything passes MAX_BLOCK_SIZE / 60, so the
result matches what it would've before.

Tests use a number equal to the number of transactions where necessary,
to ensure that they're never rejected when blocksizesize isn't being tested.

dcousens commented:
> is more amenable to any block size change proposal

@jtimon can you summarize why?   #6526 was relatively concise already,  and manually diffing that against this doesn't sound like a great use of time.

jtimon commented:
EDIT (@dcousens is going to review this):

TODO (proper description):

Compared to #6526 , in order of importance:

- [ ] This doesn't undo costly refactor work (5970  ).
- [ ] 
- [ ] 
- [ ] 
- [ ] 
- [ ]

jtimon commented:
> @jtimon can you summarize why? #6526 was relatively concise already, and manually diffing that against this doesn't sound like a great use of time.

In summary this doesn't destroy consensus/consensus.h, this doesn't s/Params().GetConsensus()/consensusParams/ when it's not necessary and (because the function depends on CChainParams, see #XXXX)...that's about it I believe.

jtimon commented:
@dcousens I edited my previous answer.

jtimon commented:
Force pushed with a fix in ```diff --git a/src/test/test_bitcoin.cpp b/src/test/test_bitcoin.cpp``` (for those who found the time to fetch a). ```CBaseChainParams::REGTEST``` was being used more than once in src/test/test_bitcoin.cpp and that caused problems in #6382 .

dcousens commented on src/main.cpp:
```diff
@@ -2591,7 +2595,10 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
-    if (block.vtx.empty() || block.vtx.size() > MAX_BLOCK_SIZE || ::GetSerializeSize(block, SER_NETWORK, PROTOCOL_VERSION) > MAX_BLOCK_SIZE)
+    uint64_t nMaxBlockSize = MaxBlockSize(consensusParams);
+    uint64_t nMaxBlockSigops = MaxBlockSigops(consensusParams);
+    uint64_t nMaxTxSize = consensusParams.nMaxTxSize;
```
Why was this pulled out as a variable?

jtimon replied:
No particular reason, ask @theuni

dcousens commented on src/consensus/consensus.h:
```diff
@@ -6,11 +6,38 @@
+/**
+ * Context-independent CTransaction validity checks
+ */
+bool CheckTx(const CTransaction& tx, CValidationState& state, uint64_t nMaxTxSize);
```
Why not just pass `consensusParams` as you have done for the other two functions?
Might as well be consistent.

dcousens commented on src/consensus/consensus.h:
```diff
@@ -6,11 +6,38 @@
+ * Check_ means checking everything possible with the data provided.
+ * Verify_ means all data provided was enough for this level and its "consensus-verified".
+ */
+namespace Consensus {
```
Why is there no indentation for this namespace block?

jtimon replied:
Because we don't indent namespaces: https://github.com/bitcoin/bitcoin/blob/master/src/.clang-format#L31

dcousens replied:
OK,  thanks for the heads up :+1:

dcousens commented:
utACK

jtimon commented:
Added a squashme commit that should solve @dcousens' nits (ie ```Consensus::CheckTx()``` gets ```const Consensus::Params& consensusParams``` instead of ```uint64_t nMaxTxSize```).

jtimon commented:
Updated with the squashed version.

jtimon commented:
Closing in favor of #6672.

jtimon commented:
Rebased and reopened.

dcousens commented:
re-ACK,  but the constant shifting of code to different PRs is getting insane. 
I'd rather not have to ACK this again.

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -484,7 +486,7 @@ void CTxMemPool::removeCoinbaseSpends(const CCoinsViewCache *pcoins, unsigned in
                 continue;
             const CCoins *coins = pcoins->AccessCoins(txin.prevout.hash);
             if (fSanityCheck) assert(coins);
```
@jtimon Needs trivial rebase.

jtimon commented:
@MarcoFalke Thanks, rebased.

jtimon commented:
New uses of MAX_BLOCK_SIZE/MaxBlockSize() were introduced in #6622.

jtimon commented:
Rebased with a tiny fix that would have break the build (if this has been merged without the fix).
I haven't squashed the new commits (from what was contained and reviewed in #6672) to facilitate re-review.

dcousens commented:
re-ACK, only did a once over though

jtimon commented:
Rebased and squashed. 
@theuni @laanwj @sipa @gmaxwell @jgarzik @gavinandresen Does this have any chance to get into 0.12 as it is?
I think it is disruptive enough that if we decide to do so, it would be good to make it happen asap.
Otherwise I would like to close it as an independent PR. It is included in both #6382 and #5970 anyway so it won't be lost, but having it synchronized in both branches plus an independent one is more annoying than it needs to be. BIP102, for example, can be maintained on top of #6382 instead.

But at the same time, if it is unlikely that it gets merged for 0.12 as it is, It would be very nice to at least create another one with the 2 or 3 commits that are suitable for 0.12 and keep maintaining the rest just as part of #6382 and #5970, at least until 0.12 is forked.

sipa commented:
utACK as-is for 0.12.

This doesn't move much code and is unlikely to conflict with other PRs apart from a few simple variable renamings. I do hope that we'll soon move the functions now declared inside consensus/*.h to their respective consensus/*.cpp, though.

jtimon commented:
Sure, I was thinking about leaving CheckTransaction/Consensus::CheckTx declared in main.h if we were going for the 2/3 commits option and that was a problem.

btcdrak commented:
I agree with @sipa, it doesn't look too disruptive as-is.

jtimon commented:
Needed trivial rebase.

MarcoFalke commented:
utACK 82b2443

btcdrak commented:
utACK

jtimon commented:
Rebased(6)

dcousens commented:
Re-ACK

jtimon commented:
Rebased(7)
Like I said the only disruptive commit is the last one and is the one that is requiring rebase all the time. Prediction: Next time this PR requires rebase, it will probably be due to the last commit and https://github.com/bitcoin/bitcoin/compare/master...jtimon:incomplete-consensus-blocksize-0.12.99 will remain without conflicts. The third commit is somewhat disruptive too, so I may be wrong, but if I am right, I will just drop the last commit, sorry @theuni

jtimon commented:
@gmaxwell 
> Bitcoin will be able to move forward with these increases when improvements and understanding render their risks widely acceptable relative to the risks of not deploying them. In Bitcoin Core we should keep patches ready to implement them as the need and the will arises, to keep the basic software engineering from being the limiting factor.

This would make it much easier to maintain those ready "patches" (even the simple bip102 has rapidly bitrot). If ithad been merged for 0.12 like was supposed to be the plan (according to sipa and btcdrak comments), that would have also make those pat hes easier to backport to 0.12.
Many people including you reviewed this as part of the closed #6672 , only a few small changes required for rebase has been added.
I'm sorry if I sound repetitive but I can't really understand what is holding this for so, so long.

dcousens commented:
This was ACK'd for `0.12` a while ago,  it should still be able to make it in right?

MarcoFalke commented:
Yes, if this get's merged in master, at least the first three commits should be ~~backported~~ (can be merged as is).

devrandom commented:
ACK

jgarzik commented:
Structure seems reasonable, with the caveat that MaxBlockSize() and MaxBlockSigOps() will inevitably be modified depending on what state they examine.

jtimon commented:
Yeah, when I changed the variable in the Consensus::Params struct (what @cfields had initially) I wasn't sure on whether to use block.nHeight or header.mediaTime(), certainly not header.nTime. If we can finally agree on that somewhere I don't mind to change it.

jgarzik commented:
ut ACK, for the record

jtimon commented:
Replaced by #7238 (a subset of this) with @rustyrussell 's nit fixed.
